{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7d414565",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import matplotlib as plt\n",
    "import torch as pt\n",
    "import csv\n",
    "import pickle #to save notebook at sessions\n",
    "\n",
    "\n",
    "#from Bojar lab format\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.model_selection import RandomizedSearchCV\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn.metrics import log_loss\n",
    "\n",
    "#set path for pickles to be saved in\n",
    "pickle_path = '/Users/erikazhang/Dropbox (MIT)/20.440 Biological Networks/project/python pickles/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ea50a98",
   "metadata": {},
   "source": [
    "# Machine Learning Workflow\n",
    "\n",
    "1. Random Forest Classifier on all T-cells (just glyco vs glyco+genes identified in 1\n",
    "- Random Forest on all T-cells w normalized data\n",
    "2. Random Forest Classifier on T-cell subsets\n",
    "3. Identify genes whose expression levels correlate with glycogene expression levels (some correlation algorithm), repeat 1 & 2 with enhanced glycogene set\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "852c2d86",
   "metadata": {},
   "source": [
    "## 1. Random Forest Clustering using all T-cells and raw data\n",
    "to build model(S) --> build on all T-cells, and then by T-cell subtypes\n",
    "https://github.com/BojarLab/scGlycomics_b16_branching/blob/main/Random%20Forest%20-%20Apr%208%202022%20-%20RQ.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f2ecfee4",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Load glycosorted RAW dataframe from pickle, saved from glycogene filtering raw.ipynb\n",
    "Dataframes have:\n",
    "- gene names as columns (243 glycogenes) + columns for type, biotin count, PHA-L score =246 columns total\n",
    "- cell barcodes as row index, for cells identified as TILs via ProjecTILs package in R\n",
    "'''\n",
    "pickle_in = open(pickle_path + \"glycosorted_TIL.pkl\",\"rb\")\n",
    "glycosorted_TIL = pickle.load(pickle_in)\n",
    "\n",
    "pickle_in = open(pickle_path + \"glycosorted_LN.pkl\",\"rb\")\n",
    "glycosorted_LN = pickle.load(pickle_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "9586d3e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make copies of dataframes to not accidentally change original df\n",
    "glycoTIL = glycosorted_TIL.copy()\n",
    "glycoLN = glycosorted_LN.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e469295",
   "metadata": {},
   "source": [
    "### Split dataset into training, validation, and test set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ed3ab9e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Generate training, validation, and test set from original full df using scikit learn \n",
    "TIL ver.\n",
    "'''\n",
    "#y: PHA-L score array\n",
    "y = glycoTIL['PHA-L'].values \n",
    "\n",
    "#X: glycogene transcript data array\n",
    "x = glycoTIL.iloc[:, :-3].values\n",
    "\n",
    "# Split training, validation and test set\n",
    "x_train_val, x_test, y_train_val, y_test = train_test_split(\n",
    "    x, y, test_size=0.1, random_state=342, stratify=y)\n",
    "\n",
    "x_train, x_val, y_train, y_val = train_test_split(\n",
    "    x_train_val, y_train_val, test_size=0.2, random_state=42, stratify=y_train_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "969b5b82",
   "metadata": {},
   "source": [
    "### Random Forest classifier model training (all from Bojar lab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "4c1cf85e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters for grid search\n",
    "\n",
    "# Number of trees in random forest\n",
    "n_estimators = [int(x) for x in np.arange(200, 800, step=100)]\n",
    "# Maximum number of levels in tree\n",
    "max_depth = [int(x) for x in np.arange(10, 50, step=10)]\n",
    "max_depth.append(None)\n",
    "# Minimum number of samples required to split a node\n",
    "min_samples_split = [2, 5]\n",
    "# Minimum number of samples required at each leaf node\n",
    "min_samples_leaf = [1, 2]\n",
    "# Create the random grid\n",
    "random_grid = {'n_estimators': n_estimators,\n",
    "               'max_depth': max_depth,\n",
    "               'min_samples_split': min_samples_split,\n",
    "               'min_samples_leaf': min_samples_leaf,\n",
    "               }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "236d4b3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 5 folds for each of 20 candidates, totalling 100 fits\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/t_/h82q6gld28zc219dv9bkpl2r0000gq/T/ipykernel_42900/886025376.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m                                   n_iter = 20, cv = 5, verbose=5, random_state=42, n_jobs = -1)\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0mmodel_random\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/sklearn/model_selection/_search.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, X, y, groups, **fit_params)\u001b[0m\n\u001b[1;32m    889\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresults\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    890\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 891\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_run_search\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mevaluate_candidates\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    892\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    893\u001b[0m             \u001b[0;31m# multimetric is determined here because in the case of a callable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/sklearn/model_selection/_search.py\u001b[0m in \u001b[0;36m_run_search\u001b[0;34m(self, evaluate_candidates)\u001b[0m\n\u001b[1;32m   1764\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_run_search\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mevaluate_candidates\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1765\u001b[0m         \u001b[0;34m\"\"\"Search n_iter candidates from param_distributions\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1766\u001b[0;31m         evaluate_candidates(\n\u001b[0m\u001b[1;32m   1767\u001b[0m             ParameterSampler(\n\u001b[1;32m   1768\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparam_distributions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mn_iter\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrandom_state\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrandom_state\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/sklearn/model_selection/_search.py\u001b[0m in \u001b[0;36mevaluate_candidates\u001b[0;34m(candidate_params, cv, more_results)\u001b[0m\n\u001b[1;32m    836\u001b[0m                     )\n\u001b[1;32m    837\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 838\u001b[0;31m                 out = parallel(\n\u001b[0m\u001b[1;32m    839\u001b[0m                     delayed(_fit_and_score)(\n\u001b[1;32m    840\u001b[0m                         \u001b[0mclone\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbase_estimator\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m   1054\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1055\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_backend\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mretrieval_context\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1056\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mretrieve\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1057\u001b[0m             \u001b[0;31m# Make sure that we get a last message telling us we are done\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1058\u001b[0m             \u001b[0melapsed_time\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_start_time\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\u001b[0m in \u001b[0;36mretrieve\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    933\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    934\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_backend\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'supports_timeout'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 935\u001b[0;31m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_output\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjob\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    936\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    937\u001b[0m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_output\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjob\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\u001b[0m in \u001b[0;36mwrap_future_result\u001b[0;34m(future, timeout)\u001b[0m\n\u001b[1;32m    540\u001b[0m         AsyncResults.get from multiprocessing.\"\"\"\n\u001b[1;32m    541\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 542\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfuture\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    543\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mCfTimeoutError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    544\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mTimeoutError\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/concurrent/futures/_base.py\u001b[0m in \u001b[0;36mresult\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    439\u001b[0m                     \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__get_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    440\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 441\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_condition\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    442\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    443\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_state\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mCANCELLED\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mCANCELLED_AND_NOTIFIED\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    310\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m    \u001b[0;31m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    311\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 312\u001b[0;31m                 \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    313\u001b[0m                 \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    314\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "''''Use RandomSearchCV to optimize hyperparameters'''\n",
    "#takes a while to run!!\n",
    "model = RandomForestClassifier()\n",
    "\n",
    "model_random = RandomizedSearchCV(estimator = model, param_distributions = random_grid, \n",
    "                                  n_iter = 20, cv = 5, verbose=5, random_state=42, n_jobs = -1)\n",
    "\n",
    "model_random.fit(x_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1cd1930",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Return the best estimator\n",
    "TILmodel_all_raw = model_random.best_estimator_\n",
    "TILmodel_all_raw.get_params()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f185be2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save pickle \n",
    "with open(pickle_path + 'TILmodel_all_raw.pkl', 'wb') as f:\n",
    "    pickle.dump(TILmodel_all_raw, f)\n",
    "\n",
    "f.close()\n",
    "\n",
    "# open via: \n",
    "# #load updated df from pickle\n",
    "# pickle_in = open(\"pickle_path + TILmodel_all_raw.pkl\",\"rb\")\n",
    "# TILmodel_all_raw = pickle.load(pickle_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1812681d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def model_evaluation(model, x, y):\n",
    "    print(f\"Accuracy for 'PHA-L high' class: {100*(model.score(x[y==1], y[y==1])):>4f}%\")\n",
    "    print(f\"Accuracy for 'PHA-L low' class: {100*(model.score(x[y==0], y[y==0])):>4f}%\")\n",
    "    print(f\"Overall accuracy: {100*(model.score(x, y)):>4f}%\")\n",
    "\n",
    "    model_predict = model.predict(x)\n",
    "    model_predict_prob = model.predict_proba(x)\n",
    "\n",
    "    print(f\"Average loss: {log_loss(y, model_predict_prob):>4f}\")\n",
    "    print(f\"ROC Curve AUC: {roc_auc_score(y, model_predict):>4f}\")\n",
    "    print(f\"F1 score: {f1_score(y, model_predict):>4f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0756d484",
   "metadata": {},
   "outputs": [],
   "source": [
    "#training set\n",
    "model_evaluation(TILmodel_all_raw, x_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c564755",
   "metadata": {},
   "outputs": [],
   "source": [
    "#validation \n",
    "model_evaluation(TILmodel_all_raw, x_val, y_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efd88efa",
   "metadata": {},
   "source": [
    "## 2B. Random Forest Classifier with LN/TIL combined T-cells\n",
    "Train models with identified subtypes\n",
    "https://github.com/BojarLab/scGlycomics_b16_branching/blob/main/Model%20training%20-%20Apr%2010%202022%20-%20RQ.ipynb\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4dc47cb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Load glycosorted RAW dataframe from pickle, saved from glycogene filtering raw.ipynb\n",
    "Dataframes have:\n",
    "- gene names as columns (243 glycogenes) + columns for type, biotin count, PHA-L score =246 columns total\n",
    "- cell barcodes as row index, for cells identified as TILs via ProjecTILs package in R\n",
    "'''\n",
    "pickle_in = open(pickle_path + \"glycosorted_TIL.pkl\",\"rb\")\n",
    "glycosorted_TIL = pickle.load(pickle_in)\n",
    "\n",
    "pickle_in = open(pickle_path + \"glycosorted_LN.pkl\",\"rb\")\n",
    "glycosorted_LN = pickle.load(pickle_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "54a3568e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make copies of dataframes to not accidentally change original df\n",
    "glycoTIL = glycosorted_TIL.copy()\n",
    "glycoLN = glycosorted_LN.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e2f6fc32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(316, 10595)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glycoLN.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "881c5333",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(316, 9824)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glycoTIL.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4d4e7808",
   "metadata": {},
   "outputs": [
    {
     "ename": "InvalidIndexError",
     "evalue": "Reindexing only valid with uniquely valued Index objects",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mInvalidIndexError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/t_/h82q6gld28zc219dv9bkpl2r0000gq/T/ipykernel_42900/2884265229.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcombo\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mglycoTIL\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mglycoLN\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/util/_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    309\u001b[0m                     \u001b[0mstacklevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstacklevel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    310\u001b[0m                 )\n\u001b[0;32m--> 311\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    313\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/concat.py\u001b[0m in \u001b[0;36mconcat\u001b[0;34m(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)\u001b[0m\n\u001b[1;32m    358\u001b[0m     )\n\u001b[1;32m    359\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 360\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    361\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    362\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/concat.py\u001b[0m in \u001b[0;36mget_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    589\u001b[0m                     \u001b[0mobj_labels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    590\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mnew_labels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mequals\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj_labels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 591\u001b[0;31m                         \u001b[0mindexers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobj_labels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew_labels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    592\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    593\u001b[0m                 \u001b[0mmgrs_indexers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_mgr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexers\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_indexer\u001b[0;34m(self, target, method, limit, tolerance)\u001b[0m\n\u001b[1;32m   3727\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3728\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_index_as_unique\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3729\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mInvalidIndexError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_requires_unique_msg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3730\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3731\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtarget\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mInvalidIndexError\u001b[0m: Reindexing only valid with uniquely valued Index objects"
     ]
    }
   ],
   "source": [
    "combo = pd.concat([glycoTIL, glycoLN])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c0e57ab7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TACAGTGGTTCAACCA-1</th>\n",
       "      <th>ACATACGAGTGCCATT-1</th>\n",
       "      <th>TGTCCCATCGGCTTGG-1</th>\n",
       "      <th>ACTGTCCAGGATGGTC-1</th>\n",
       "      <th>GAACATCTCTGCTGTC-1</th>\n",
       "      <th>CATCAAGAGACAATAC-1</th>\n",
       "      <th>GAAACTCCAGATTGCT-1</th>\n",
       "      <th>CAGCATACAGCCTATA-1</th>\n",
       "      <th>TAGAGCTAGGAGTCTG-1</th>\n",
       "      <th>GGGATGAGTCAGGACA-1</th>\n",
       "      <th>...</th>\n",
       "      <th>GGCGTGTAGAAGGGTA-1</th>\n",
       "      <th>CGGAGTCAGCGCTTAT-1</th>\n",
       "      <th>GGACAAGAGTGGTAAT-1</th>\n",
       "      <th>CTGATAGAGAATGTGT-1</th>\n",
       "      <th>AGCGTATGTTCGAATC-1</th>\n",
       "      <th>AGATTGCCACGAAGCA-1</th>\n",
       "      <th>GGCGTGTAGGGATCTG-1</th>\n",
       "      <th>CGTCTACTCAGCCTAA-1</th>\n",
       "      <th>GCTGGGTCACCGGAAA-1</th>\n",
       "      <th>CTGAAGTGTCTCATCC-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ahsa1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Api5</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Atp6v1e1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bcap31</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cops6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dsel</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Glce</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Type</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CD8_NaiveLike</td>\n",
       "      <td>CD8_NaiveLike</td>\n",
       "      <td>CD8_NaiveLike</td>\n",
       "      <td>CD8_NaiveLike</td>\n",
       "      <td>CD8_NaiveLike</td>\n",
       "      <td>CD8_NaiveLike</td>\n",
       "      <td>CD8_NaiveLike</td>\n",
       "      <td>CD8_NaiveLike</td>\n",
       "      <td>CD8_NaiveLike</td>\n",
       "      <td>CD8_NaiveLike</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Biotin_hash</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHA-L</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>632 rows Ã— 20275 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            TACAGTGGTTCAACCA-1 ACATACGAGTGCCATT-1 TGTCCCATCGGCTTGG-1  \\\n",
       "Ahsa1                        0                  0                  0   \n",
       "Api5                         1                  2                  0   \n",
       "Atp6v1e1                     1                  0                  0   \n",
       "Bcap31                       1                  0                  1   \n",
       "Cops6                        0                  0                  0   \n",
       "...                        ...                ...                ...   \n",
       "Dsel                       NaN                NaN                NaN   \n",
       "Glce                       NaN                NaN                NaN   \n",
       "Type                       NaN                NaN                NaN   \n",
       "Biotin_hash                NaN                NaN                NaN   \n",
       "PHA-L                      NaN                NaN                NaN   \n",
       "\n",
       "            ACTGTCCAGGATGGTC-1 GAACATCTCTGCTGTC-1 CATCAAGAGACAATAC-1  \\\n",
       "Ahsa1                        0                  3                  2   \n",
       "Api5                         1                  4                  2   \n",
       "Atp6v1e1                     0                  2                  1   \n",
       "Bcap31                       1                  7                  2   \n",
       "Cops6                        0                  3                  5   \n",
       "...                        ...                ...                ...   \n",
       "Dsel                       NaN                NaN                NaN   \n",
       "Glce                       NaN                NaN                NaN   \n",
       "Type                       NaN                NaN                NaN   \n",
       "Biotin_hash                NaN                NaN                NaN   \n",
       "PHA-L                      NaN                NaN                NaN   \n",
       "\n",
       "            GAAACTCCAGATTGCT-1 CAGCATACAGCCTATA-1 TAGAGCTAGGAGTCTG-1  \\\n",
       "Ahsa1                        2                  0                  4   \n",
       "Api5                         0                  0                  0   \n",
       "Atp6v1e1                     0                  0                  2   \n",
       "Bcap31                       1                  1                  5   \n",
       "Cops6                        2                  0                  0   \n",
       "...                        ...                ...                ...   \n",
       "Dsel                       NaN                NaN                NaN   \n",
       "Glce                       NaN                NaN                NaN   \n",
       "Type                       NaN                NaN                NaN   \n",
       "Biotin_hash                NaN                NaN                NaN   \n",
       "PHA-L                      NaN                NaN                NaN   \n",
       "\n",
       "            GGGATGAGTCAGGACA-1  ... GGCGTGTAGAAGGGTA-1 CGGAGTCAGCGCTTAT-1  \\\n",
       "Ahsa1                        0  ...                NaN                NaN   \n",
       "Api5                         0  ...                NaN                NaN   \n",
       "Atp6v1e1                     0  ...                NaN                NaN   \n",
       "Bcap31                       2  ...                NaN                NaN   \n",
       "Cops6                        0  ...                NaN                NaN   \n",
       "...                        ...  ...                ...                ...   \n",
       "Dsel                       NaN  ...                  0                  0   \n",
       "Glce                       NaN  ...                  0                  0   \n",
       "Type                       NaN  ...      CD8_NaiveLike      CD8_NaiveLike   \n",
       "Biotin_hash                NaN  ...                 20                 19   \n",
       "PHA-L                      NaN  ...                0.0                0.0   \n",
       "\n",
       "            GGACAAGAGTGGTAAT-1 CTGATAGAGAATGTGT-1 AGCGTATGTTCGAATC-1  \\\n",
       "Ahsa1                      NaN                NaN                NaN   \n",
       "Api5                       NaN                NaN                NaN   \n",
       "Atp6v1e1                   NaN                NaN                NaN   \n",
       "Bcap31                     NaN                NaN                NaN   \n",
       "Cops6                      NaN                NaN                NaN   \n",
       "...                        ...                ...                ...   \n",
       "Dsel                         0                  0                  0   \n",
       "Glce                         0                  0                  0   \n",
       "Type             CD8_NaiveLike      CD8_NaiveLike      CD8_NaiveLike   \n",
       "Biotin_hash                 18                 18                 16   \n",
       "PHA-L                      0.0                0.0                0.0   \n",
       "\n",
       "            AGATTGCCACGAAGCA-1 GGCGTGTAGGGATCTG-1 CGTCTACTCAGCCTAA-1  \\\n",
       "Ahsa1                      NaN                NaN                NaN   \n",
       "Api5                       NaN                NaN                NaN   \n",
       "Atp6v1e1                   NaN                NaN                NaN   \n",
       "Bcap31                     NaN                NaN                NaN   \n",
       "Cops6                      NaN                NaN                NaN   \n",
       "...                        ...                ...                ...   \n",
       "Dsel                         0                  0                  0   \n",
       "Glce                         0                  0                  0   \n",
       "Type             CD8_NaiveLike      CD8_NaiveLike      CD8_NaiveLike   \n",
       "Biotin_hash                 16                 14                 10   \n",
       "PHA-L                      0.0                0.0                0.0   \n",
       "\n",
       "            GCTGGGTCACCGGAAA-1 CTGAAGTGTCTCATCC-1  \n",
       "Ahsa1                      NaN                NaN  \n",
       "Api5                       NaN                NaN  \n",
       "Atp6v1e1                   NaN                NaN  \n",
       "Bcap31                     NaN                NaN  \n",
       "Cops6                      NaN                NaN  \n",
       "...                        ...                ...  \n",
       "Dsel                         0                  0  \n",
       "Glce                         0                  0  \n",
       "Type             CD8_NaiveLike      CD8_NaiveLike  \n",
       "Biotin_hash                 10                  6  \n",
       "PHA-L                      0.0                0.0  \n",
       "\n",
       "[632 rows x 20275 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bdf0487",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Generate training, validation, and test set from original full df using scikit learn \n",
    "TIL ver.\n",
    "'''\n",
    "#y: PHA-L score array\n",
    "y = glycoTIL['PHA-L'].values \n",
    "\n",
    "#X: glycogene transcript data array\n",
    "x = glycoTIL.iloc[:, :-3].values\n",
    "\n",
    "# Split training, validation and test set\n",
    "x_train_val, x_test, y_train_val, y_test = train_test_split(\n",
    "    x, y, test_size=0.1, random_state=342, stratify=y)\n",
    "\n",
    "x_train, x_val, y_train, y_val = train_test_split(\n",
    "    x_train_val, y_train_val, test_size=0.2, random_state=42, stratify=y_train_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e6667fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters for grid search\n",
    "\n",
    "# Number of trees in random forest\n",
    "n_estimators = [int(x) for x in np.arange(200, 800, step=100)]\n",
    "# Maximum number of levels in tree\n",
    "max_depth = [int(x) for x in np.arange(10, 50, step=10)]\n",
    "max_depth.append(None)\n",
    "# Minimum number of samples required to split a node\n",
    "min_samples_split = [2, 5]\n",
    "# Minimum number of samples required at each leaf node\n",
    "min_samples_leaf = [1, 2]\n",
    "# Create the random grid\n",
    "random_grid = {'n_estimators': n_estimators,\n",
    "               'max_depth': max_depth,\n",
    "               'min_samples_split': min_samples_split,\n",
    "               'min_samples_leaf': min_samples_leaf,\n",
    "               }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "495c7d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "''''Use RandomSearchCV to optimize hyperparameters'''\n",
    "#takes a while to run!!\n",
    "model = RandomForestClassifier()\n",
    "\n",
    "combomodel_all_raw = RandomizedSearchCV(estimator = model, param_distributions = random_grid, \n",
    "                                  n_iter = 20, cv = 5, verbose=5, random_state=42, n_jobs = -1)\n",
    "\n",
    "combomodel_all_raw.fit(x_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82a404ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save pickle \n",
    "with open(pickle_path + 'combomodel_all_raw.pkl', 'wb') as f:\n",
    "    pickle.dump(combomodel_all_raw, f)\n",
    "\n",
    "f.close()\n",
    "\n",
    "# open via: \n",
    "# #load updated df from pickle\n",
    "# pickle_in = open(\"pickle_path + combomodel_all_raw.pkl\",\"rb\")\n",
    "# combomodel_all_raw = pickle.load(pickle_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd90d14d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Return the best estimator\n",
    "combomodel_all_raw = model_random.best_estimator_\n",
    "combomodel_all_raw.get_params()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3b09906",
   "metadata": {},
   "outputs": [],
   "source": [
    "#training set\n",
    "model_evaluation(TILmodel_all_raw, x_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc0cc4ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "#validation \n",
    "model_evaluation(TILmodel_all_raw, x_val, y_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3dd21f3",
   "metadata": {},
   "source": [
    "## 3. Random Forest Classifier with T-cell subsets\n",
    "Train models with identified subtypes\n",
    "https://github.com/BojarLab/scGlycomics_b16_branching/blob/main/Model%20training%20-%20Apr%2010%202022%20-%20RQ.ipynb\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c9c4cfdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "##load updated df from pickle\n",
    "pickle_in = open(pickle_path + \"TILtcell_dfs.pkl\",\"rb\")\n",
    "TILtcell_dfs = pickle.load(pickle_in)\n",
    "\n",
    "pickle_in = open(pickle_path + \"LNtcell_dfs.pkl\",\"rb\")\n",
    "LNtcell_dfs = pickle.load(pickle_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e433dde3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#extracting all dataframes from df dictionary\n",
    "CD8_NaiveLike_df= TILtcell_dfs['CD8_NaiveLike_df']\n",
    "CD8_EffectorMemory_df = TILtcell_dfs['CD8_EffectorMemory_df']\n",
    "Th1_df = TILtcell_dfs['Th1_df']\n",
    "CD8_EarlyActiv_df = TILtcell_dfs['CD8_EarlyActiv_df']\n",
    "Treg_df = TILtcell_dfs['Treg_df']\n",
    "CD8_Tex_df = TILtcell_dfs['CD8_Tex_df']\n",
    "CD4_NaiveLike_df = TILtcell_dfs['CD4_NaiveLike_df']\n",
    "Tfh_df = TILtcell_dfs['Tfh_df']\n",
    "CD8_Tpex_df = TILtcell_dfs['CD8_Tpex_df']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
