{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "053145e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import matplotlib as plt\n",
    "import torch as pt\n",
    "import csv\n",
    "import pickle #to save notebook at sessions\n",
    "\n",
    "\n",
    "#from Bojar lab format\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.model_selection import RandomizedSearchCV\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn.metrics import log_loss\n",
    "from sklearn.metrics import accuracy_score\n",
    "\n",
    "#set path for pickles to be saved in\n",
    "pickle_path = '/Users/erikazhang/Dropbox (MIT)/20.440 Biological Networks/project/python pickles/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53774d6a",
   "metadata": {},
   "source": [
    "# Random Forest Clustering for combo, separated by different normalization methods (T-cells from LN and TILs) and compare the different normalization methods"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "febabc45",
   "metadata": {},
   "source": [
    "## Set up:\n",
    "- open pickle\n",
    "- define model_evaluation\n",
    "- set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b3f8d59b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open raw combo dataframe\n",
    "#load updated df from pickle\n",
    "pickle_in = open(pickle_path +\"combo_raw.pkl\",\"rb\")\n",
    "combo_raw = pickle.load(pickle_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8fe557e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def model_evaluation(model, x, y):\n",
    "    print(f\"Accuracy for 'PHA-L high' class: {100*(model.score(x[y==1], y[y==1])):>4f}%\")\n",
    "    print(f\"Accuracy for 'PHA-L low' class: {100*(model.score(x[y==0], y[y==0])):>4f}%\")\n",
    "    print(f\"Overall accuracy: {100*(model.score(x, y)):>4f}%\")\n",
    "    high_accuracy = 100*(model.score(x[y==1], y[y==1]))\n",
    "    low_accuracy = 100*(model.score(x[y==0], y[y==0]))\n",
    "    overall_accuracy = 100*(model.score(x, y))\n",
    "        \n",
    "    model_predict = model.predict(x)\n",
    "    model_predict_prob = model.predict_proba(x)\n",
    "\n",
    "    print(f\"Average loss: {log_loss(y, model_predict_prob):>4f}\")\n",
    "    print(f\"ROC Curve AUC: {roc_auc_score(y, model_predict):>4f}\")\n",
    "    print(f\"F1 score: {f1_score(y, model_predict):>4f}\")\n",
    "    return high_accuracy, low_accuracy, overall_accuracy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "558b6c51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters for grid search\n",
    "\n",
    "# Number of trees in random forest\n",
    "n_estimators = [int(x) for x in np.arange(200, 800, step=100)]\n",
    "# Maximum number of levels in tree\n",
    "max_depth = [int(x) for x in np.arange(10, 50, step=10)]\n",
    "max_depth.append(None)\n",
    "# Minimum number of samples required to split a node\n",
    "min_samples_split = [2, 5]\n",
    "# Minimum number of samples required at each leaf node\n",
    "min_samples_leaf = [1, 2]\n",
    "# Create the random grid\n",
    "random_grid = {'n_estimators': n_estimators,\n",
    "               'max_depth': max_depth,\n",
    "               'min_samples_split': min_samples_split,\n",
    "               'min_samples_leaf': min_samples_leaf,\n",
    "               }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "27c6d9c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from sklearn.preprocessing import RobustScaler\n",
    "from sklearn.preprocessing import LabelEncoder\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e312de8e",
   "metadata": {},
   "source": [
    "## Split up combo by T-cell subtype (make dictionary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e80d1657",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Now want to split up glycosorted dataframe into sub-dataframes for\n",
    "each of the different T-cell subtypes. TO do this, this code block takes in the glycosorted\n",
    "dataframe data as the input and spits out a dictionaryt containing:\n",
    "- t-cell type as key \n",
    "- matrix containing raw expression data for dataframe as values\n",
    "\n",
    "'''\n",
    "# Get list of T-cell subtypes for which to make a dataframe for TILs\n",
    "tcell_subtypes = combo_raw['Type'].unique()\n",
    "\n",
    "#make dataframe names for later access\n",
    "df_names = [i+'_df' for i in list(tcell_subtypes)]\n",
    "\n",
    "# Make copy of original df just in case\n",
    "split_df = combo_raw.copy()\n",
    "\n",
    "combotcell_dfs = {}\n",
    "\n",
    "# Make separate dataframe containing data for each t-cell subtype\n",
    "for cell_type, name in zip(tcell_subtypes, df_names):\n",
    "    combotcell_dfs[name] = split_df[split_df['Type'] == cell_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10b90981",
   "metadata": {},
   "source": [
    "### Standard scaler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e8e97d8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/t_/h82q6gld28zc219dv9bkpl2r0000gq/T/ipykernel_86758/3903640621.py:24: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  toy_df2[cols_to_standardize] = scaler.fit_transform(toy_df2[cols_to_standardize])\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'LabelEncoder' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/t_/h82q6gld28zc219dv9bkpl2r0000gq/T/ipykernel_86758/3903640621.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m     \u001b[0;31m#encode all values as numbers not anything else\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 46\u001b[0;31m     \u001b[0mencoder\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mLabelEncoder\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     47\u001b[0m     \u001b[0my_train\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mencoder\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit_transform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m     \u001b[0my_val\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mencoder\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit_transform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_val\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'LabelEncoder' is not defined"
     ]
    }
   ],
   "source": [
    "# make dataframe to store subtype's overall accuracy scores \n",
    "cols = {\"Cell type\": [], \"PHA-L high accuracy\": [], \"PHA-L low accuracy\": [], \n",
    "        \"Overall accuracy\": [], 'Dimensions':[], 'Top 5 features':[]}\n",
    "standard_accuracy = pd.DataFrame(columns = cols)\n",
    "\n",
    "# for loop to add values for each t-cell subtype into dataframe\n",
    "subtype_names = list(combotcell_dfs.keys())\n",
    "\n",
    "for name in subtype_names:\n",
    "    '''\n",
    "    Generate training, validation, and test set from original full df using scikit learn \n",
    "    TIL ver.\n",
    "    '''\n",
    "    ## generate dataframe from dictionary key\n",
    "    sub_df = combotcell_dfs[name]\n",
    "    \n",
    "    ###NORMALIZE DATA using robust scaler\n",
    "    \n",
    "    ### NORMALIZE COLUMNS ###\n",
    "    toy_df = sub_df.transpose()\n",
    "    toy_df2 = toy_df.iloc[:-4] #takes out the non-gene expression columns\n",
    "    cols_to_standardize = toy_df2.columns    \n",
    "    scaler = StandardScaler()\n",
    "    toy_df2[cols_to_standardize] = scaler.fit_transform(toy_df2[cols_to_standardize])\n",
    "\n",
    "    ### ADD BACK Cell type, Biotin values, and L-PHA scores ###\n",
    "    scores_to_append = toy_df.iloc[-4:]\n",
    "    df = pd.concat([toy_df2, scores_to_append])\n",
    "    df = df.transpose()\n",
    "    \n",
    "    ## start random forest here\n",
    "    #y: PHA-L score array\n",
    "    y = df['PHA-L'].values \n",
    "\n",
    "    #X: glycogene transcript data array\n",
    "    x = df.iloc[:, :-4].values\n",
    "\n",
    "    # Split training, validation and test set\n",
    "    x_train_val, x_test, y_train_val, y_test = train_test_split(\n",
    "        x, y, test_size=0.1, random_state=42, stratify=y)\n",
    "\n",
    "    x_train, x_val, y_train, y_val = train_test_split(\n",
    "        x_train_val, y_train_val, test_size=0.2, random_state=42, stratify=y_train_val)\n",
    "    \n",
    "    #encode all values as numbers not anything else \n",
    "    encoder = LabelEncoder()\n",
    "    y_train = encoder.fit_transform(y_train)\n",
    "    y_val = encoder.fit_transform(y_val)\n",
    "    \n",
    "    ''''Use RandomSearchCV to optimize hyperparameters'''\n",
    "    \n",
    "    #Generate model!!\n",
    "    model = RandomForestClassifier()\n",
    "\n",
    "    model_random = RandomizedSearchCV(estimator = model, param_distributions = random_grid, \n",
    "                                      n_iter = 20, cv = 5, verbose=5, random_state=42, n_jobs = -1)\n",
    "\n",
    "    model_random.fit(x_train, y_train)\n",
    "    \n",
    "    # Return the best estimator\n",
    "    TILmodel_sub_robust = model_random.best_estimator_\n",
    "    \n",
    "    #train model\n",
    "    model_evaluation(TILmodel_sub_robust, x_train, y_train)\n",
    "    \n",
    "    #test model\n",
    "    high, low, total = model_evaluation(TILmodel_sub_robust, x_val, y_val)\n",
    "    \n",
    "    #extract top 5 important genes and append to dataframe \n",
    "    feature_importances = pd.DataFrame({'feature': sub_df.columns[:-3], \n",
    "                                        'importance': TILmodel_sub_robust.feature_importances_})\n",
    "\n",
    "    # Sort the dataframe by importance score in descending order\n",
    "    feature_importances = feature_importances.sort_values('importance', ascending=False)\n",
    "    feature_importances['combo'] = feature_importances['feature'] + ': '+ feature_importances['importance'].round(4).astype(str)\n",
    "\n",
    "    # make new column that combines feature name with its importance value, get top 5 as string to add to dictionary\n",
    "    top5_df = feature_importances.head()\n",
    "    top5 = str(list(top5_df['combo']))\n",
    "    \n",
    "    #save all the info into dataframe\n",
    "    #get dimension \n",
    "    dim = sub_df.shape\n",
    "    standard_accuracy.loc[len(subtype_accuracy.index)] = [name, high, low, total, dim, top5] "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "853f042e",
   "metadata": {},
   "source": [
    "### MinMax scaler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b0d3aa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make dataframe to store subtype's overall accuracy scores \n",
    "cols = {\"Cell type\": [], \"PHA-L high accuracy\": [], \"PHA-L low accuracy\": [], \n",
    "        \"Overall accuracy\": [], 'Dimensions':[], 'Top 5 features':[]}\n",
    "minmax_accuracy = pd.DataFrame(columns = cols)\n",
    "\n",
    "# for loop to add values for each t-cell subtype into dataframe\n",
    "subtype_names = list(combotcell_dfs.keys())\n",
    "\n",
    "for name in subtype_names:\n",
    "    '''\n",
    "    Generate training, validation, and test set from original full df using scikit learn \n",
    "    TIL ver.\n",
    "    '''\n",
    "    ## generate dataframe from dictionary key\n",
    "    sub_df = combotcell_dfs[name]\n",
    "    \n",
    "    ###NORMALIZE DATA using robust scaler\n",
    "    \n",
    "    ### NORMALIZE COLUMNS ###\n",
    "    toy_df = sub_df.transpose()\n",
    "    toy_df2 = toy_df.iloc[:-4] #takes out the non-gene expression columns\n",
    "    cols_to_standardize = toy_df2.columns    \n",
    "    scaler = MinMaxScaler()\n",
    "    toy_df2[cols_to_standardize] = scaler.fit_transform(toy_df2[cols_to_standardize])\n",
    "\n",
    "    ### ADD BACK Cell type, Biotin values, and L-PHA scores ###\n",
    "    scores_to_append = toy_df.iloc[-4:]\n",
    "    df = pd.concat([toy_df2, scores_to_append])\n",
    "    df = df.transpose()\n",
    "    \n",
    "    ## start random forest here\n",
    "    #y: PHA-L score array\n",
    "    y = df['PHA-L'].values \n",
    "\n",
    "    #X: glycogene transcript data array\n",
    "    x = df.iloc[:, :-4].values\n",
    "\n",
    "    # Split training, validation and test set\n",
    "    x_train_val, x_test, y_train_val, y_test = train_test_split(\n",
    "        x, y, test_size=0.1, random_state=42, stratify=y)\n",
    "\n",
    "    x_train, x_val, y_train, y_val = train_test_split(\n",
    "        x_train_val, y_train_val, test_size=0.2, random_state=42, stratify=y_train_val)\n",
    "    \n",
    "    #encode all values as numbers not anything else \n",
    "    encoder = LabelEncoder()\n",
    "    y_train = encoder.fit_transform(y_train)\n",
    "    y_val = encoder.fit_transform(y_val)\n",
    "    \n",
    "    ''''Use RandomSearchCV to optimize hyperparameters'''\n",
    "    \n",
    "    #Generate model!!\n",
    "    model = RandomForestClassifier()\n",
    "\n",
    "    model_random = RandomizedSearchCV(estimator = model, param_distributions = random_grid, \n",
    "                                      n_iter = 20, cv = 5, verbose=5, random_state=42, n_jobs = -1)\n",
    "\n",
    "    model_random.fit(x_train, y_train)\n",
    "    \n",
    "    # Return the best estimator\n",
    "    TILmodel_sub_robust = model_random.best_estimator_\n",
    "    \n",
    "    #train model\n",
    "    model_evaluation(TILmodel_sub_robust, x_train, y_train)\n",
    "    \n",
    "    #test model\n",
    "    high, low, total = model_evaluation(TILmodel_sub_robust, x_val, y_val)\n",
    "    \n",
    "    #extract top 5 important genes and append to dataframe \n",
    "    feature_importances = pd.DataFrame({'feature': sub_df.columns[:-3], \n",
    "                                        'importance': TILmodel_sub_robust.feature_importances_})\n",
    "\n",
    "    # Sort the dataframe by importance score in descending order\n",
    "    feature_importances = feature_importances.sort_values('importance', ascending=False)\n",
    "    feature_importances['combo'] = feature_importances['feature'] + ': '+ feature_importances['importance'].round(4).astype(str)\n",
    "\n",
    "    # make new column that combines feature name with its importance value, get top 5 as string to add to dictionary\n",
    "    top5_df = feature_importances.head()\n",
    "    top5 = str(list(top5_df['combo']))\n",
    "    \n",
    "    #save all the info into dataframe\n",
    "    #get dimension \n",
    "    dim = sub_df.shape\n",
    "    minmax_accuracy.loc[len(subtype_accuracy.index)] = [name, high, low, total, dim, top5] \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "021e2714",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make dataframe to store subtype's overall accuracy scores \n",
    "cols = {\"Cell type\": [], \"PHA-L high accuracy\": [], \"PHA-L low accuracy\": [], \n",
    "        \"Overall accuracy\": [], 'Dimensions':[], 'Top 5 features':[]}\n",
    "robust_accuracy = pd.DataFrame(columns = cols)\n",
    "\n",
    "# for loop to add values for each t-cell subtype into dataframe\n",
    "subtype_names = list(combotcell_dfs.keys())\n",
    "\n",
    "for name in subtype_names:\n",
    "    '''\n",
    "    Generate training, validation, and test set from original full df using scikit learn \n",
    "    TIL ver.\n",
    "    '''\n",
    "    ## generate dataframe from dictionary key\n",
    "    sub_df = combotcell_dfs[name]\n",
    "    \n",
    "    ###NORMALIZE DATA using robust scaler\n",
    "    \n",
    "    ### NORMALIZE COLUMNS ###\n",
    "    toy_df = sub_df.transpose()\n",
    "    toy_df2 = toy_df.iloc[:-4] #takes out the non-gene expression columns\n",
    "    cols_to_standardize = toy_df2.columns    \n",
    "    scaler = RobustScaler()\n",
    "    toy_df2[cols_to_standardize] = scaler.fit_transform(toy_df2[cols_to_standardize])\n",
    "\n",
    "    ### ADD BACK Cell type, Biotin values, and L-PHA scores ###\n",
    "    scores_to_append = toy_df.iloc[-4:]\n",
    "    df = pd.concat([toy_df2, scores_to_append])\n",
    "    df = df.transpose()\n",
    "    \n",
    "    ## start random forest here\n",
    "    #y: PHA-L score array\n",
    "    y = df['PHA-L'].values \n",
    "\n",
    "    #X: glycogene transcript data array\n",
    "    x = df.iloc[:, :-4].values\n",
    "\n",
    "    # Split training, validation and test set\n",
    "    x_train_val, x_test, y_train_val, y_test = train_test_split(\n",
    "        x, y, test_size=0.1, random_state=42, stratify=y)\n",
    "\n",
    "    x_train, x_val, y_train, y_val = train_test_split(\n",
    "        x_train_val, y_train_val, test_size=0.2, random_state=42, stratify=y_train_val)\n",
    "    \n",
    "    #encode all values as numbers not anything else \n",
    "    encoder = LabelEncoder()\n",
    "    y_train = encoder.fit_transform(y_train)\n",
    "    y_val = encoder.fit_transform(y_val)\n",
    "    \n",
    "    ''''Use RandomSearchCV to optimize hyperparameters'''\n",
    "    \n",
    "    #Generate model!!\n",
    "    model = RandomForestClassifier()\n",
    "\n",
    "    model_random = RandomizedSearchCV(estimator = model, param_distributions = random_grid, \n",
    "                                      n_iter = 20, cv = 5, verbose=5, random_state=42, n_jobs = -1)\n",
    "\n",
    "    model_random.fit(x_train, y_train)\n",
    "    \n",
    "    # Return the best estimator\n",
    "    TILmodel_sub_robust = model_random.best_estimator_\n",
    "    \n",
    "    #train model\n",
    "    model_evaluation(TILmodel_sub_robust, x_train, y_train)\n",
    "    \n",
    "    #test model\n",
    "    high, low, total = model_evaluation(TILmodel_sub_robust, x_val, y_val)\n",
    "    \n",
    "    #extract top 5 important genes and append to dataframe \n",
    "    feature_importances = pd.DataFrame({'feature': sub_df.columns[:-3], \n",
    "                                        'importance': TILmodel_sub_robust.feature_importances_})\n",
    "\n",
    "    # Sort the dataframe by importance score in descending order\n",
    "    feature_importances = feature_importances.sort_values('importance', ascending=False)\n",
    "    feature_importances['combo'] = feature_importances['feature'] + ': '+ feature_importances['importance'].round(4).astype(str)\n",
    "\n",
    "    # make new column that combines feature name with its importance value, get top 5 as string to add to dictionary\n",
    "    top5_df = feature_importances.head()\n",
    "    top5 = str(list(top5_df['combo']))\n",
    "    \n",
    "    #save all the info into dataframe\n",
    "    #get dimension \n",
    "    dim = sub_df.shape\n",
    "    robust_accuracy.loc[len(subtype_accuracy.index)] = [name, high, low, total, dim, top5] \n",
    "    \n",
    "    \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
